File: android/app/src/main/kotlin/com/example/ble_proximity_bridge/StatusPayloadSupplier.kt
package com.example.ble_proximity_bridge

import io.heraldprox.herald.sensor.Device
import io.heraldprox.herald.sensor.datatype.Data
import io.heraldprox.herald.sensor.datatype.PayloadData
import io.heraldprox.herald.sensor.datatype.PayloadTimestamp
import io.heraldprox.herald.sensor.datatype.UInt32
import io.heraldprox.herald.sensor.datatype.UInt64
import io.heraldprox.herald.sensor.datatype.UInt8
import io.heraldprox.herald.sensor.payload.DefaultPayloadDataSupplier

/**
 * Encodes the payload we broadcast and decodes payloads from peers. The
 * same 16 bytes also travel in goodbye/hello frames, which is why the
 * decoders take Herald's base [Data] type.
 *
 * Wire format (16 bytes, little-endian, must match iOS exactly):
 *
 * ```
 * offset 0-7   UInt64  peer id
 * offset 8     UInt8   status code
 * offset 9     UInt8   color index
 * offset 10    UInt8   device kind (0 = iOS, 1 = Android)
 * offset 11    UInt8   flags (bit 0 = offline / goodbye)
 * offset 12-15 UInt32  protocol version
 * ```
 *
 * The offline flag matters because BLE has no built-in goodbye. Decoders
 * are defensive: bad payloads are dropped, not allowed to crash anything.
 */
class StatusPayloadSupplier(peerId: Long) : DefaultPayloadDataSupplier() {

    companion object {
        const val DEVICE_KIND_ANDROID = 1
        const val PROTOCOL_VERSION = 2L
        const val FLAG_OFFLINE = 0x01

        private const val OFFSET_PEER_ID = 0
        private const val OFFSET_STATUS = 8
        private const val OFFSET_COLOR = 9
        private const val OFFSET_DEVICE_KIND = 10
        private const val OFFSET_FLAGS = 11
        private const val OFFSET_VERSION = 12
        private const val PAYLOAD_SIZE = 16

        fun peerIdFrom(data: Data?): Long? {
            if (data == null || data.value.size < PAYLOAD_SIZE) return null
            return data.uint64(OFFSET_PEER_ID)?.value
        }

        fun statusFrom(data: Data?): Int? {
            if (data == null || data.value.size < PAYLOAD_SIZE) return null
            return data.subdata(OFFSET_STATUS)?.uint8(0)?.value
        }

        fun colorFrom(data: Data?): Int? {
            if (data == null || data.value.size < PAYLOAD_SIZE) return null
            return data.subdata(OFFSET_COLOR)?.uint8(0)?.value
        }

        fun deviceKindFrom(data: Data?): Int? {
            if (data == null || data.value.size < PAYLOAD_SIZE) return null
            return data.subdata(OFFSET_DEVICE_KIND)?.uint8(0)?.value
        }

        fun isOfflineFrom(data: Data?): Boolean? {
            if (data == null || data.value.size < PAYLOAD_SIZE) return null
            val flags = data.subdata(OFFSET_FLAGS)?.uint8(0)?.value ?: return null
            return flags and FLAG_OFFLINE != 0
        }

        fun versionFrom(data: Data?): Long? {
            if (data == null || data.value.size < PAYLOAD_SIZE) return null
            return data.subdata(OFFSET_VERSION)?.uint32(0)?.value
        }
    }

    // Herald reads this on BLE threads while the platform thread writes
    // updates, so all access goes through the lock.
    private val lock = Any()
    private var peerId: Long = peerId
    private var status: Int = 0
    private var color: Int = 0
    private var offline: Boolean = false

    fun updatePeerId(peerId: Long) {
        synchronized(lock) {
            this.peerId = peerId
        }
    }

    fun update(status: Int, color: Int) {
        synchronized(lock) {
            this.status = status
            this.color = color
        }
    }

    /** While offline, every payload we serve carries the goodbye flag. */
    fun setOffline(offline: Boolean) {
        synchronized(lock) {
            this.offline = offline
        }
    }

    /** Returns the current payload as a goodbye/hello frame. */
    fun currentFrame(): PayloadData = payload(PayloadTimestamp(), null)

    override fun payload(timestamp: PayloadTimestamp, device: Device?): PayloadData {
        val id: Long
        val statusNow: Int
        val colorNow: Int
        val flags: Int
        synchronized(lock) {
            id = peerId
            statusNow = status
            colorNow = color
            flags = if (offline) FLAG_OFFLINE else 0
        }

        return PayloadData().apply {
            append(UInt64(id))
            append(UInt8(statusNow))
            append(UInt8(colorNow))
            append(UInt8(DEVICE_KIND_ANDROID))
            append(UInt8(flags))
            append(UInt32(PROTOCOL_VERSION))
        }
    }
}

File: ios/Runner/StatusPayloadSupplier.swift
import Foundation
import Herald

/// Encodes the payload we broadcast and decodes payloads from peers. The
/// same 16 bytes also travel in goodbye/hello frames, which is why the
/// decoders take raw `Data`.
///
/// Wire format (16 bytes, little-endian, must match the Android encoder):
///
/// ```
/// offset 0-7   UInt64  peer id
/// offset 8     UInt8   status code
/// offset 9     UInt8   color index
/// offset 10    UInt8   device kind (0 = iOS, 1 = Android)
/// offset 11    UInt8   flags (bit 0 = offline / goodbye)
/// offset 12-15 UInt32  protocol version
/// ```
///
/// The offline flag exists because BLE has no goodbye of its own. Peers that
/// are still connected keep reading our cached payload after we stop, and
/// the flag is how they find out we're gone.
///
/// The decoders never throw or crash on bad input; a short or garbled
/// payload is just dropped.
final class StatusPayloadSupplier: PayloadDataSupplier {

  static let payloadSize = 16
  static let deviceKindIOS: UInt8 = 0
  static let protocolVersion: UInt32 = 2
  static let flagOffline: UInt8 = 0x01

  // Herald reads from its BLE queues while the platform thread writes,
  // so everything goes through the lock.
  private let lock = NSLock()
  private var peerId: UInt64
  private var status: UInt8 = 0
  private var color: UInt8 = 0
  private var offline = false

  init(peerId: Int) {
    // Bit-pattern conversion so a negative id round-trips instead of
    // trapping — the Android encoder (Kotlin Long) has the same semantics.
    self.peerId = UInt64(bitPattern: Int64(peerId))
  }

  func update(status: Int, color: Int) {
    lock.lock()
    defer { lock.unlock() }
    self.status = UInt8(clamping: status)
    self.color = UInt8(clamping: color)
  }

  func updatePeerId(_ peerId: Int) {
    lock.lock()
    defer { lock.unlock() }
    self.peerId = UInt64(bitPattern: Int64(peerId))
  }

  /// While offline, every payload we serve carries the goodbye flag,
  /// including re-reads from peers whose connections outlived our stop.
  func setOffline(_ value: Bool) {
    lock.lock()
    defer { lock.unlock() }
    offline = value
  }

  // MARK: - PayloadDataSupplier

  func payload(_ timestamp: PayloadTimestamp, device: Device?) -> PayloadData? {
    lock.lock()
    let id = peerId
    let statusNow = status
    let colorNow = color
    let flags: UInt8 = offline ? Self.flagOffline : 0
    lock.unlock()

    let result = PayloadData()
    result.append(id)  // bytes 0-7
    result.append(statusNow)  // byte 8
    result.append(colorNow)  // byte 9
    result.append(Self.deviceKindIOS)  // byte 10
    result.append(flags)  // byte 11
    result.append(Self.protocolVersion)  // bytes 12-15
    return result
  }

  /// The current payload as raw bytes, for immediate-send frames.
  func currentFrame() -> Data? {
    payload(PayloadTimestamp(), device: nil)?.data
  }

  // MARK: - Decoders (raw Data, shared by payload reads and immediate-send)

  static func peerId(from data: Data) -> Int? {
    guard data.count >= payloadSize, let value = data.uint64(0) else {
      return nil
    }
    // Bit-pattern conversion, not Int(value): an id with the top bit set
    // (any device could broadcast one on our service UUID) must decode as
    // a negative Int — matching Kotlin's signed Long — instead of trapping.
    return Int(bitPattern: UInt(value))
  }

  static func status(from data: Data) -> Int? {
    guard data.count >= payloadSize,
      let value = data.advanced(by: 8).uint8(0)
    else { return nil }
    return Int(value)
  }

  static func color(from data: Data) -> Int? {
    guard data.count >= payloadSize,
      let value = data.advanced(by: 9).uint8(0)
    else { return nil }
    return Int(value)
  }

  static func deviceKind(from data: Data) -> Int? {
    guard data.count >= payloadSize,
      let value = data.advanced(by: 10).uint8(0)
    else { return nil }
    return Int(value)
  }

  static func isOffline(from data: Data) -> Bool? {
    guard data.count >= payloadSize,
      let flags = data.advanced(by: 11).uint8(0)
    else { return nil }
    return flags & flagOffline != 0
  }

  static func version(from data: Data) -> Int? {
    guard data.count >= payloadSize,
      let value = data.advanced(by: 12).uint32(0)
    else { return nil }
    return Int(value)
  }

  // MARK: - Decoder conveniences for Herald PayloadData

  static func peerId(from payload: PayloadData) -> Int? { peerId(from: payload.data) }
  static func status(from payload: PayloadData) -> Int? { status(from: payload.data) }
  static func color(from payload: PayloadData) -> Int? { color(from: payload.data) }
  static func deviceKind(from payload: PayloadData) -> Int? { deviceKind(from: payload.data) }
  static func isOffline(from payload: PayloadData) -> Bool? { isOffline(from: payload.data) }
  static func version(from payload: PayloadData) -> Int? { version(from: payload.data) }
}
